diff --git a/Makefile b/Makefile
index fd6e888..79f83b8 100644
--- a/Makefile
+++ b/Makefile
@@ -5,24 +5,24 @@ ifdef D
 	OPT=
 else
 	DEBUG=
-	OPT=-Ofast
+	OPT=
 endif
 
 ifdef NH
 	ARCH=
 else
-	ARCH=-msse4.2 -D__SSE4_2_
+	ARCH=
 endif
 
 ifdef P
 	PROFILE=-pg -no-pie # for bug in gprof.
 endif
 
-CXX = g++ -std=c++11
-CC = gcc -std=gnu11
-LD= g++ -std=c++11
+CXX = g++-11 -std=c++11
+CC = gcc-11 -std=gnu11
+LD= g++-11 -std=c++11
 
-LOC_INCLUDE=include
+LOC_INCLUDE=/project/archive-index-data/seiler/nature_hibf/repositories/squeakr/include
 LOC_SRC=src
 OBJDIR=obj
 
@@ -30,8 +30,8 @@ CXXFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -I$(LOC_INCLUDE)
 
 CFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -I$(LOC_INCLUDE)
 
-LDFLAGS += $(DEBUG) $(PROFILE) $(OPT) -lpthread -lboost_system \
--lboost_thread -lm -lbz2 -lz -lrt
+LDFLAGS += $(DEBUG) $(PROFILE) $(OPT) -lpthread \
+-lm -lbz2 -lz -lrt -Llib/ -lboost_system -lboost_thread
 
 #
 # declaration of dependencies
diff --git a/src/kmer.cc b/src/kmer.cc
index 0e282cb..fd4412e 100644
--- a/src/kmer.cc
+++ b/src/kmer.cc
@@ -10,7 +10,6 @@
 
 #include <fstream>
 #include "kmer.h"
-#include <stdlib.h>
 #include <iostream>
 
 /*return the integer representation of the base */
